lib/internal/util/comparisons.js

```diff
@@ -770,11 +876,28 @@ function sparseArrayEquiv(a, b, mode, memos, i) {
   return true;
 }
 
-function objEquiv(a, b, mode, keys2, memos, iterationType) {
+function objEquiv(a, b, mode, keys1, keys2, memos, iterationType) {
   // The pair must have equivalent values for every corresponding key.
   if (keys2.length > 0) {
-    for (const key of keys2) {
-      if (!innerDeepEqual(a[key], b[key], mode, memos)) {
+    let i = 0;
+    // Ordered keys
+    if (keys1 !== undefined) {
+      for (; i < keys2.length; i++) {
+        const key = keys2[i];
+        if (keys1[i] !== key) {
+          break;
+        }
+        if (!innerDeepEqual(a[key], b[key], mode, memos)) {
+          return false;
+        }
+      }
+    }
+    // Unordered keys
+    for (; i < keys2.length; i++) {
+      const key = keys2[i];
+      const descriptor = ObjectGetOwnPropertyDescriptor(a, key);
```

Comment:
This warrants a comment as to why this is here instead of `ObjectPropertyIsEnumerable`.